jerry-core/ecma/builtin-objects/ecma-builtin-promise.c

```diff
@@ -577,10 +577,19 @@ ecma_builtin_promise_race_or_all (ecma_value_t this_arg, /**< 'this' argument */
       || ecma_get_object_type (ecma_get_object_from_value (array)) != ECMA_OBJECT_TYPE_ARRAY)
   {
     ecma_raise_type_error (ECMA_ERR_MSG ("Second argument is not an array."));
+    ecma_value_t current_error_value = JERRY_CONTEXT (error_value);
     ret = ecma_builtin_promise_reject_abrupt (capability);
-    ecma_free_value (JERRY_CONTEXT (error_value));
-    ecma_free_value (capability);
 
+    if (JERRY_CONTEXT (error_value) == current_error_value)
```

Comment:
Why is this if needed? If they are equal or not you must free `current_error_value`. Am I miss something?

Reply:
The reason of this `if` that `ecma_raise_type_error` always sets the `JERRY_CONTEXT(error_value)` (and this error must be freed like before) but `ecma_builtin_promise_reject_abrupt` can override this value. In this case the correct behavior is to keep the overriden error_value and free the previously raised TypeError.

Reply:
Why dont we postpone the error construction after the `ecma_builtin_promise_reject_abrupt` call?

Reply:
On the necessity of the `if`, I go with @LaszloLango here. If `JERRY_CONTEXT (error_value) == current_error_value` (this is the predicate of the if) then `ecma_free_value (JERRY_CONTEXT (error_value))` (which is in the true branch of the if) should be rewritable to `ecma_free_value (current_error_value)`. But then the true and false branches of the if become equivalent, making all this simply a `ecma_free_value (current_error_value);`.

All which means that @zherczeg 's suggestion is most probably right: swapping `ecma_raise_type_error` and `ecma_builtin_promise_reject_abrupt` _should_ give an equivalent and much simpler solution.

Reply:
I @akosthekiss @LaszloLango Finally I've understood your thoughts, and this `if` in unnecessary, because the raised TypeError __must__ be freed anyway, but we have to make sure that we free the correct error. (That's why the we store the `JERRY_CONTEXT (error_value)` in a local variable, to prevent `ecma_builtin_promise_reject_abrupt` from override the `JERRY_CONTEXT (error_value)`.)

@zherczeg Swapping the `ecma_raise_type_error` and `ecma_builtin_promise_reject_abrupt` is not a way to go, because the raised TypeError is going to be the reason of the promise rejection, that's why this order is needed.

I've updated the PR, IMO this solution is much simpler then the previous one.